library.go

```diff
@@ -0,0 +1,158 @@
+	// not implemented on this specific implementation.
+	ErrNotImplemented = errors.NewKind("not implemented")
+	// ErrModeNotSupported is returned in the case of a request to open a
+	// repository with a Mode not supported.
+	ErrModeNotSupported = errors.NewKind("repository mode %q not supported")
+	// ErrLocationNotExists when a Location is requested and can't be found.
+	ErrLocationNotExists = errors.NewKind("location %s not exists")
+	// ErrLibraryNotExists when a Library is requested and can't be found.
+	ErrLibraryNotExists = errors.NewKind("library %s not exists")
+	// ErrRepositoryExists an error returned on a request of Init on a location
+	// with a repository with this RepositoryID already exists.
+	ErrRepositoryExists = errors.NewKind("repository %s already exists")
+	// ErrRepositoryNotExists when a Repository is requested and can't be found.
+	ErrRepositoryNotExists = errors.NewKind("repository %s not exists")
+	// ErrNonTransactional returned when Repository.Commit is called on a
+	// repository that not support transactions.
+	ErrNonTransactional = errors.NewKind("non transactional repository")
+)
+
+// RepositoryID represents a Repository identifier, these IDs regularly are
+// based on a http or git remore URL, but can be based on any other concept.
+type RepositoryID string
+
+// NewRepositoryID returns a new RepositoryID based on a given endpoint.
+// Eg.: git@github.com:src-d/go-borges becomes github.com/src-d/go-borges.git
+func NewRepositoryID(endpoint string) (RepositoryID, error) {
+	e, err := transport.NewEndpoint(endpoint)
+	if err != nil {
+		return "", err
+	}
+
+	if !strings.HasSuffix(e.Path, ".git") {
+		e.Path += ".git"
+	}
+
+	return RepositoryID(path.Join(e.Host, e.Path)), nil
+}
+
+func (id RepositoryID) String() string {
+	return string(id)
+}
+
+// Mode is the different modes to open a Repository.
+type Mode int
+
+const (
+	// RWMode allows to perform read and write operations over a repository.
+	RWMode Mode = iota
+	// ReadOnlyMode allows only read-only operations over a repository.
+	ReadOnlyMode
+)
+
+// Repository interface represents a git.Repository, with information about
+// how it was open and where is located. Also provides Transactional
+// capabilities through the method Commit.
+type Repository interface {
+	// ID returns the RepositoryID.
+	ID() RepositoryID
+	// LocationID returns the LocationID from the Location where it was retrieved.
+	LocationID() LocationID
+	// Mode returns the Mode how it was opened.
+	Mode() Mode
+	// Commit persists all the write operations done since was open, if the
+	// repository doesn't provide Transactional capabilities should return
+	// ErrNonTransactional.
+	Commit() error
+	// Close closes the repository, if the repository was opened in transactional
+	// Mode, will delete any write operation pending to be written.
+	Close() error
+	// R returns the git.Repository.
+	R() *git.Repository
```

Comment:
We could be more explicit with this method's name and avoid code like `r.R()` for library users.

Reply:
This is exactly the goal, be short. Otherwise, we will have tons of `repository.Repository()`

Reply:
`repo.Get()`?